### PR TITLE
feat: add no_std support

### PR DESCRIPTION
**Pull request by Rqnsom**
Some of the changes are taken from the another deim/bcs fork:
https://github.com/zefchain/bcs

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -1,23 +1,25 @@
 [package]
 name = "bcs"
 version = "0.1.4"
-authors = ["Diem <opensource@diem.com>"]
 description = "Binary Canonical Serialization (BCS)"
-repository = "https://github.com/diem/bcs"
-homepage = "https://diem.com"
+repository = "https://github.com/zefchain/bcs"
```
Comment (asmie) [inline Cargo.toml]: So we've forked from diem repository but setting it to zefchain - maybe we should form directly from zefchain if we are using their version?

**Review by asmie — APPROVED** [review]
LGTM 👍 

We could probably divide this commit into smaller, single feature-related ones (even within one PR) as there are some groups of changes made.
1) serializer changes - all those things that change `W` to `Vec<u8>` and remove std::io::Write.
2) Error handling - the changes with `fmt::Display` for Error.
3) Other no_std changes - like `std` to `core` replace, Clippy arithmetic etc.

File: Cargo.toml (hunk 1 above)
Comment (Rqnsom) [reply Cargo.toml]: gooood catch! I'll revert this part and leave a comment about it or something, hmm

**Comment by Rqnsom** [comment]
> We could probably divide this commit into smaller, single feature-related ones (even within one PR) as there are some groups of changes made.

I do agree, but that would take a lot of our time.

**Pull request merged**